src/auth.js

```diff
@@ -1,197 +1,161 @@
- *    `signature = sign(hash(envelope))`)
- * @param {string} publicKey  the public identity that signed this envelope
+ * // Assume signPayloadCallback is a well-formed signing callback.
+ * //
+ * // It might, for example, pop up a modal from a browser extension, send the
+ * // transaction to a third-party service for signing, or just do simple
+ * // signing via Keypair like it does here:
+ * function signPayloadCallback(payload) {
+ *    const signature = signer.sign(hash(payload.toXDR());
+ *    return [ signature, signer.publicKey() ];
+ * }
  *
- * @returns {xdr.SorobanAuthorizationEntry}
+ * function multiPartyAuth(
+ *    server: Server,
+ *    // assume this involves multi-party auth
+ *    tx: Transaction,
+ * ) {
+ *    return server
+ *      .simulateTransaction(tx)
+ *      .then((simResult) => {
+ *          tx.operations[0].auth.map(entry =>
+ *            authorizeEntry(
+ *              entry,
+ *              signPayloadCallback,
+ *              currentLedger + 1000,
+ *              Networks.FUTURENET);
+ *          ));
  *
- * @throws {Error} if `verify(hash(envelope), signature, publicKey)` does not
- *    pass, meaning one of the arguments was not passed or built correctly
- * @throws {TypeError} if the envelope does not hold an
- *    {@link xdr.HashIdPreimageSorobanAuthorization} instance
+ *          return server.prepareTransaction(tx, Networks.FUTURENET, simResult);
+ *      })
+ *      .then((preppedTx) => {
+ *        preppedTx.sign(source);
+ *        return server.sendTransaction(preppedTx);
+ *      });
+ * }
  */
-export function buildAuthEntry(envelope, signature, publicKey) {
-  // ensure this identity signed this envelope correctly
+export async function authorizeEntry(
+  entry,
+  signingMethod,
+  validUntil,
+  networkPassphrase = Networks.FUTURENET
+) {
+  // no-op
   if (
-    !Keypair.fromPublicKey(publicKey).verify(hash(envelope.toXDR()), signature)
+    entry.credentials().switch() !==
+    xdr.SorobanCredentialsType.sorobanCredentialsAddress()
   ) {
-    throw new Error(`signature does not match envelope or identity`);
+    return entry;
   }
 
-  if (
-    envelope.switch() !== xdr.EnvelopeType.envelopeTypeSorobanAuthorization()
-  ) {
-    throw new TypeError(
-      `expected sorobanAuthorization envelope, got ${envelope.switch().name}`
+  /** @type {xdr.SorobanAddressCredentials} */
+  const addrAuth = entry.credentials().address();
+  addrAuth.signatureExpirationLedger(validUntil);
+
+  const networkId = hash(Buffer.from(networkPassphrase));
+  const preimage = xdr.HashIdPreimage.envelopeTypeSorobanAuthorization(
+    new xdr.HashIdPreimageSorobanAuthorization({
+      networkId,
+      nonce: addrAuth.nonce(),
+      invocation: entry.rootInvocation(),
+      signatureExpirationLedger: addrAuth.signatureExpirationLedger()
+    })
+  );
+  const payload = hash(preimage.toXDR());
+
+  let signature;
+  let publicKey;
+  if (signingMethod instanceof Keypair) {
```

Comment:
Side note: you could actually make this fn more general by switching the if/else blocks, like:
```js
let sign = signingMethod;
if (typeof signingMethod !== 'function') {
  // Would handle Keypair, or any object with a `.sign`
  // method, like: `{sign: (payload) => bytes}`
  sign = signingMethod.sign;
}
const signature = sign(payload);
// etc...

Reply:
ooh great call, done!